src/Akka.Streams.Kafka/Stages/Consumers/CommitCollectorStage.cs

```diff
@@ -0,0 +1,167 @@
+﻿using System;
+using Akka.Streams.Kafka.Messages;
+using Akka.Streams.Kafka.Settings;
+using Akka.Streams.Stage;
+
+namespace Akka.Streams.Kafka.Stages.Consumers;
+
+/// <summary>
+/// INTERNAL API
+///
+/// Combined stage for committing incoming offsets in batches.
+/// </summary>
+internal sealed class CommitCollectorStage : GraphStage<FlowShape<ICommittable, ICommittableOffsetBatch>>
+{
+    private readonly CommitterSettings _settings;
+
+    public Inlet<ICommittable> In { get; } = new("FlowIn");
+    public Outlet<ICommittableOffsetBatch> Out { get; } = new("FlowOut");
+
+    public CommitCollectorStage(CommitterSettings settings)
+    {
+        _settings = settings;
+        Shape = new FlowShape<ICommittable, ICommittableOffsetBatch>(In, Out);
+    }
+
+    public override FlowShape<ICommittable, ICommittableOffsetBatch> Shape { get; }
+
+    protected override GraphStageLogic CreateLogic(Attributes inheritedAttributes) =>
+        new CommitCollectorStageLogic(Shape, this, inheritedAttributes);
+
+    private class CommitCollectorStageLogic : TimerGraphStageLogic
+    {
+        private const string CommitNow = "FlowStageCommit";
+
+        public CommitCollectorStageLogic(Shape shape, CommitCollectorStage stage, Attributes inheritedAttributes) :
+            base(shape)
+        {
+            Stage = stage;
+            InheritedAttributes = inheritedAttributes;
+            ObservationLogic = new CommitObservationLogic(stage._settings);
+            SetHandler(Stage.Out, OnPull);
+            SetHandler(Stage.In, OnPush, OnUpstreamFinish, OnUpstreamFailure);
+        }
+
+        private void OnUpstreamFailure(Exception ex)
+        {
+            Log.Debug(ex, "OnUpstreamFailure with exception with {0}", ObservationLogic.OffsetBatch);
+            if (ActiveBatchInProgress)
+            {
+                // TODO: probably need to do some base-class work or something to avoid interface abuse here
+                ((CommittableOffsetBatch)ObservationLogic.OffsetBatch).TellCommitEmergency();
+                ObservationLogic.OffsetBatch = CommittableOffsetBatch.Empty;
+            }
+            FailStage(ex);
+        }
+
+        private void OnUpstreamFinish()
+        {
+            if (ActiveBatchInProgress)
+            {
+                Log.Debug("PushDownStream triggered by {0}, outstanding batch {1}", CommitTrigger.UpstreamFinish.Instance,
+                    ObservationLogic.OffsetBatch);
+                Emit(Stage.Out, ObservationLogic.OffsetBatch);
+            }
+            CompleteStage();
+        }
+
+        private void OnPush()
+        {
+            var offset = Grab(Stage.In);
+            Log.Debug("Consuming offset {0}", offset);
+            if (ObservationLogic.UpdateBatch(offset))
```

Comment:
If the latest offset we've received triggers an update to our batch (newer `Offset`) then we'll check our commit triggers and update the batch accordingly.